pkg/datastore/postgres.go

```diff
@@ -353,3 +422,72 @@ func (ds *Datastore) Find(ctx context.Context, filter map[string]interface{}, re
 
 	return err
 }
+
+// Get the history entity for given id and the given point in time
+// returns NotFoundError if no entity can be found
+func (ds *Datastore) GetHistory(ctx context.Context, id string, at time.Time, ve VersionedJSONEntity) error {
+	jsonField := ve.JSONField()
+	tableName := historyTablename(ve.TableName())
+	_, ok := ds.types[jsonField]
+	if !ok {
+		return fmt.Errorf("type:%s is not registered", jsonField)
+	}
+	q := ds.sb.Select(jsonField).From(tableName).Where(squirrel.And{
+		squirrel.Eq{
+			"id": id,
+		},
+		squirrel.LtOrEq{
+			"created_at": at,
+		},
+	}).OrderByClause("created_at DESC").Limit(1)
+
+	sql, _, _ := q.ToSql()
+	ds.log.Info("get", zap.String("entity", jsonField), zap.String("sql", sql), zap.String("id", id), zap.String("created_at", at.Format(time.RFC3339)))
+	rows, err := q.QueryContext(ctx)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		cerr := rows.Close()
+		if err == nil && cerr != nil {
+			err = cerr
+		}
+	}()
+	if rows.Next() {
+		return rows.Scan(ve)
+	}
+	// we have no row
+	return NewNotFoundError(fmt.Sprintf("entity of type:%s with id:%s not found", jsonField, id))
+}
+
+func (ds *Datastore) insertHistory(ve VersionedJSONEntity, op Op, runner squirrel.BaseRunner) error {
+	jsonField := ve.JSONField()
+	tableName := ve.TableName()
+	qh := ds.sb.Insert(historyTablename(tableName)).
+		SetMap(map[string]interface{}{
+			"id":         ve.GetMeta().Id,
+			"op":         op,
+			"created_at": Now(),
+			jsonField:    ve,
+		})
+	_, err := qh.RunWith(runner).Exec()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// historyTablename returns the tablename of the table-twin with historic data.
+func historyTablename(table string) string {
+	return fmt.Sprintf("%s_history", table)
+}
+
+// PbNow return
+func PbNow() *timestamp.Timestamp {
+	now, err := ptypes.TimestampProto(Now())
+	if err != nil {
+		panic(err)
```

Comment:
this is a bit harsh

Reply:
The only possibility for this error to occur is in testing when someone fails to provide a correct Now-func when overriding for testing at a defined point in time.
What is more: it is exactly what github.com/golang/protobuf@v1.3.5/ptypes/timestamp.go:103 does.
So I would suggest to leave it this way, even I also don't like panics in "harmless" funcs.

Reply:
Decided this will not happen